tensorflow/contrib/tensorrt/BUILD

```diff
@@ -67,6 +67,7 @@ tf_cuda_library(
     visibility = ["//visibility:public"],
     deps = [
         ":trt_logging",
+        ":trt_plugins",
```

Comment:
I think there will be double import problems here: the plugin library has a singleton, but will be imported twice by two different python library (trt_convert_py and trt_ops_py), so at the end there will be two different singletons. @samikama to confirm.

Reply:
After some analysis, I'm pretty sure this will cause problem:

1. the plugin factory is a global instance
2. it is linked here by trt_shape_function, which will eventually be linked into python/ops/_trt_engine_op.so which will be loaded by the python library
3. it is also linked into trt_engine_op_kernel and trt_conversion, which will eventually linked into _wrap_conversion.so which will be loaded by the python library
4. so there will be two copies of the factory, which is dangerous and will cause problems

I would suggest to make a header-only library and let trt_shape_function depend on it instead. The Logger class in the trt_logging rule also has this problem, but since it doesn't contain any data, it should be fine (but I still want to fix it later).

Reply:
I thought linux dynamic linking would be able to resolve duplicate libraries at runtime, not sure how it would perform on other os neither.
But I'm not very good with these stuff and the solution sounds straightforward. Will update and ping you.

Reply:
So currently trt_shape_function requires plugin factory to deserialize the engine; so does trt_conversion.
It does not seem clear to me how I can turn plugin factory into a header-only library to break the potential double import. It would be great if you could provide more details :)

Technically speaking, I can change trt_shape_function to use attributes to store output tensor shapes (required through TF static shape inference during segmentation) and avoid engine deserialization. 
This would work better with our optimization pass & last minute conversion during execution.

Reply:
Quick update:
We should remove the deserialization in trt_shape_function. Sami is also doing the same thing in his current PR where he's modifying the trt_engine_op. To avoid conflict I'll wait for his PR to be merged and then merge on top of that.

Reply:
My understanding is that linker will load the first copy of the factory instance and ignore the second, regardless which .so is loaded first. Then, if the one who is not linked against the plugin registration is loaded first (maybe _trt_engine_op.so), then later when it tried to find the plugin it will fail. I could be wrong though.

We can see that both .so file has GetInstance() defined as a unique global symbol.

Reply:
Thanks J, just saw your updates, that works for me. Meanwhile I'm fixing the formatting issues.

Reply:
Was digging a bit more into this. Since we will remove the code, so it's not related anymore.

I start to think this might not be an issue. Member function from an ordinary class has external linkage and could be accessed in other translation units. http://en.cppreference.com/w/cpp/language/storage_duration 
I tested it locally on with gcc. 
I have two shared libraries (lib1.so, lib2.so) all linked against a singleton library (libstg.so) . My client application could access the singleton through interfaces exposed in lib1.so & lib2.so. I think dynamic linker takes the first encounter of the getInstance and uses it for both (Windows might behave differently tho).

One side thing I caught is that I didn't guard my singleton initialization between threads. Will update that.